### PR TITLE
Add ARM64 Support

### LINKED ISSUE #1865: Add arm64 build for native components
We removed this quite some time ago since the CI used at the time didn't support arm64. As time goes by arm64 is getting more popular and we should build our native components to support arm64.

### LINKED ISSUE #2181: Add arm64 architecture to the shared store
The list at [dotnet/runtime]( https://github.com/dotnet/runtime/blob/42ef1970047ace8972e22b13e5b05a7e2be656d2/src/native/corehost/hostmisc/utils.cpp#LL201-LL209) has other values but because of Apple M1 boxes, we may want to provide `arm64` by default.

Adding all `arch` to the shared store doesn't seem scalable, ideally, we would have a way to automatically select the proper `|arch|/|tfm|` pair upon installing on the box. However, as mentioned, `arm64` may be popular enough to justify its addition even without the general mechanism.

**Temporary Workaround**: rename `$DOTNET_SHARED_STORE/x64/` to `$DOTNET_SHARED_STORE/arm64/`. The assemblies under the folder are IL only so [source instrumentations](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/docs/design.md#library-instrumentations) are expected to work without issues. The native component is not supported on arm64 yet and given that the default rule engine will fail startup if it is enabled it needs to be disabled removing or setting the environment variable `CORECLR_ENABLE_PROFILING` to `0`.

[EDIT 01: Update workaround since it is not going to work with the default rule engine checks.]

### PR DESCRIPTION
**Pull request by RassK**
## Why

Fixes #1865
Fixes #2181

Towards #2314
Towards #2453

## What

Adds ARM64 support for ubuntu, debian and alpine.
CentOS based images are **not supported** with .NET 6 and .NET 7

## Tests

Tests .NET 6, .NET 7 and .NET 8.

* arm64 ubuntu 22.04 (native platform)
* arm64 debian (container)
* arm64 alpine (container)
* arm64 nuget package tests

## Follow up

`Confluent.Kafka 1.4.0`  is not supported by arm64, is disabled temporarily due missing platform filtration capabilities.
`Confluent.Kafka 1.6.2` does not work on x64 and arm64. This seems to require separate focus.
Follow up task here > https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/3306

## Checklist

- [x] `CHANGELOG.md` is updated.
- [x] Documentation is updated.
- [x] New features are covered by tests.

### CODE REVIEW

**Comment by RassK** [comment]
Note: Docs issue, it's looking for a dockerfile from a main branch, that does not exist before merge.

File: .github/workflows/build.yml
Comment (pjanotti) [inline .github/workflows/build.yml:39]: Why this is not needed by the other architectures?

File: test/IntegrationTests/Helpers/EnvironmentTools.cs
Comment (pjanotti) [inline test/IntegrationTests/Helpers/EnvironmentTools.cs:101]: why the different casing here?

File: test/IntegrationTests/IntegrationTests.csproj
Comment (pjanotti) [inline test/IntegrationTests/IntegrationTests.csproj:7]: Is this really needed?

File: .github/workflows/build.yml
Comment (RassK) [reply .github/workflows/build.yml:39]: `actions/setup-dotnet@v4.0.0` had somekind of issues running without root privileges. I guess other runners are configured differently.

File: test/IntegrationTests/Helpers/EnvironmentTools.cs
Comment (RassK) [reply test/IntegrationTests/Helpers/EnvironmentTools.cs:101]: Not to cause (weird) issues with Visual Studio. I have no idea why they defined ARM64 all uppercase in $(AvailablePlatforms).

File: test/IntegrationTests/IntegrationTests.csproj
Comment (RassK) [reply test/IntegrationTests/IntegrationTests.csproj:7]: Yes, this is connected to _X64 and _ARM64 constant. Visual Studio is using a `$(Platform)` variable to define these constants. It also makes Visual Studio interacting with the code that is marked as `#if _X64` so as soon as you select x64 the code has intellisense.

File: .github/workflows/build.yml
Comment (RassK) [reply .github/workflows/build.yml:39]:
```
Run actions/setup-dotnet@v4.0.0
/home/runner/actions-runner/_work/_actions/actions/setup-dotnet/v4.0.0/externals/install-dotnet.sh --skip-non-versioned-files --runtime dotnet --channel LTS
mkdir: cannot create directory ‘/usr/share/dotnet’: Permission denied
```

File: .github/workflows/release-publish.yml
Comment (Kielek) [inline .github/workflows/release-publish.yml:1]: Did you have chance to test it, e.g. on your fork? Maybe you can make something like 1.4.0-alpha.1?

File: docs/README.md
Comment (Kielek) [inline docs/README.md:11]: Tracked this syntax on: https://github.com/self-actuated/docs.actuated.dev/issues/18

File: CHANGELOG.md
Comment (Kielek) [inline CHANGELOG.md:15]: I would consider putting something like - experimental support for (here and in the docs). With this, we can release ARM64 in stable packages. Without this, it should be supported only on prereleases. It applies also to REDME.md

**Review by Kielek — COMMENTED** [review]
Good job. Some minor comments.

File: .github/workflows/build-container.yml
Comment (pellared) [inline .github/workflows/build-container.yml:20]: Is there any list of what Actuated machines are available? Any docs that would help us in troubleshooting?

Comment (RassK) [reply .github/workflows/build-container.yml:20]: I think maintainers should see it? Actions -> (left side label: management) runners

Comment (pellared) [reply .github/workflows/build-container.yml:20]: I see only regular GitHub runners.

FYI related discussion: https://github.com/open-telemetry/opentelemetry-go/pull/4994#discussion_r1508641578

Comment (RassK) [reply .github/workflows/build-container.yml:20]: Yes that would have been next issue after there was more information, which runners are even available.
I was told also that they are going to remove the default label somewhere in the near? future.

File: tools/LibraryVersionsGenerator/PackageVersionDefinitions.cs
Comment (RassK) [reply tools/LibraryVersionsGenerator/PackageVersionDefinitions.cs]: Interestingly 1.6.2 causes issues also on x64 while 1.4.0 was ok 🤔 This is the first release that ships arm64 native binaries also. Might need bigger research, we might have holes in the supported range.

File: .github/workflows/build-container.yml
Comment (pellared) [reply .github/workflows/build-container.yml:20]: Look like it would be removed rather sooner than later. I suggest changing this line to 

```suggestion
          - machine: actuated-arm64-4cpu-8gb
```

https://github.com/open-telemetry/opentelemetry-go/pull/4994#discussion_r1509178904

Comment (RassK) [reply .github/workflows/build-container.yml:20]: will update within this scope then 👍

File: test/IntegrationTests/IntegrationTests.csproj
Comment (zacharycmontoya) [reply test/IntegrationTests/IntegrationTests.csproj:7]: Instead of assigning a Platform here and in the Nuke code, can we 1) compile these as AnyCPU, 2) remove the ifdefs, and 3) check the architecture dynamically when the test is run and skip the test then?

The reason I ask is that the solution file is building this project as AnyCPU for release builds but platform-specific for debug builds, and I'd like to build managed code as AnyCPU as much as possible.

Comment (RassK) [reply test/IntegrationTests/IntegrationTests.csproj:7]: The only issue with runtime check was that it has to be done in multiple places. 
1. In the test itself, need to change Theory to SkippableTheory and then skip the test if arm64
2. In the `SqlServerFixture` (or any fixture that does not work) cause this executes way before the test and fails there.

Is a bit more fragmented approach but still not that complicated to do. Personally don't have strong feelings about either of the approach.

File: .github/workflows/build.yml
Comment (RassK) [reply .github/workflows/build.yml:39]: https://github.com/actions/setup-dotnet/issues/486

File: build/Build.cs
Comment (rajkumar-rangaraj) [inline build/Build.cs:14]: nit: Provide a better clarity her - "Default is current architecture."

Comment (RassK) [reply build/Build.cs:14]: any suggestions?

File: .github/workflows/build-container.yml
Comment (RassK) [reply .github/workflows/build-container.yml]: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/3277/commits/2cbc49253876099d927e59f27c1ba9ef194c6f53

File: CHANGELOG.md
Comment (RassK) [reply CHANGELOG.md]: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/3277/commits/33c5bf20fe54381f78566fe8b0b74f13be8b8915

**Review by Kielek — APPROVED** [review]

File: test/IntegrationTests/IntegrationTests.csproj
Comment (pjanotti) [reply test/IntegrationTests/IntegrationTests.csproj:7]: I like @zacharycmontoya suggestion, but, it can come in a follow up.

**Review by pjanotti — APPROVED** [review]

File: test/IntegrationTests/IntegrationTests.csproj
Comment (RassK) [reply test/IntegrationTests/IntegrationTests.csproj:7]: I think @Kielek was currently the only one who preferred more compile time filter than runtime... which is definitely cleaner solution. But up to you... I'd prefer to do it in the current branch (while there is history to check just in case)

File: docs/README.md
Comment (RassK) [reply docs/README.md:11]: Badge is merged 👍 I'll update the docs also in the current branch tomorrow.

File: test/IntegrationTests/IntegrationTests.csproj
Comment (Kielek) [reply test/IntegrationTests/IntegrationTests.csproj:7]: From my perspective, it is just a preference. Any working options should be fine :)

File: docs/README.md
Comment (Kielek) [reply docs/README.md:11]:
```suggestion
[![Arm CI sponsored by Actuated](https://img.shields.io/badge/SA_actuated.dev-004BDD)](https://actuated.dev/)
```

Great!

File: .markdownlint.yaml
Comment (Kielek) [inline .markdownlint.yaml:19]:
```suggestion
```

This should be also not needed.

**Review by pellared — APPROVED** [review]

**Comment by Kielek** [comment]
@rajkumar-rangaraj, @zacharycmontoya, do you see any objections to merge it?

If no, I will adjust merge build requierments.

**Comment by pellared** [comment]
> @rajkumar-rangaraj, @zacharycmontoya, do you see any objections to merge it?
> 
> If no, I will adjust merge build requierments.

See https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/3307

**Review by zacharycmontoya — APPROVED** [review]

File: test/IntegrationTests/IntegrationTests.csproj
Comment (zacharycmontoya) [reply test/IntegrationTests/IntegrationTests.csproj:7]: Sounds good, it's good to merge and we can perform this as follow-up 👍🏼

**Pull request merged**
